### PR TITLE
feat: add knowledge_brain

### PR DESCRIPTION
**Pull request by chloedia**
I have made changes to the knowledge table schema and models to improve functionality and data organization. This includes adding new columns, modifying existing columns, and updating relationships. These changes will enhance the overall performance and usability of the application.

### CODE REVIEW

File: backend/core/quivr_core/models.py [hunk 1]
```diff
@@ -8,6 +7,7 @@
 from langchain_core.pydantic_v1 import BaseModel as BaseModelV1
 from langchain_core.pydantic_v1 import Field as FieldV1
 from pydantic import BaseModel
+from quivr_api.modules.knowledge.dto.inputs import KnowledgeStatus
```
Comment (AmineDiro) [inline backend/core/quivr_core/models.py:10]: quivr_api depends on quivr_core we can't have imports from quivr_api in quivr_core

File: backend/core/quivr_core/models.py [hunk 2]
```diff
@@ -83,12 +83,6 @@ class ParsedRAGChunkResponse(BaseModel):
     last_chunk: bool = False
 
 
-class KnowledgeStatus(str, Enum):
```
Comment (AmineDiro) [inline backend/core/quivr_core/models.py:86]: I think we should rollback this

File: backend/supabase/migrations/20240808141057_knowledge_fix.sql [hunk 3]
```diff
@@ -1,4 +1,10 @@
 alter table "public"."knowledge" drop column "file_sha";
 
+alter table "public"."knowledge" drop column "metadata";
+
 alter table "public"."knowledge" add column "file_sha1" text;
+
+alter table "public"."knowledge" add column "metadata" jsonb;
```
Comment (AmineDiro) [inline backend/supabase/migrations/20240808141057_knowledge_fix.sql:7]: Unnecessary changes I think. We already have the metadata column

File: backend/supabase/migrations/20240814091741_notion-sync.sql [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+alter table "public"."knowledge_brain" validate constraint "public_knowledge_brain_brain_id_fkey";
+
+alter table "public"."knowledge_brain" add constraint "public_knowledge_brain_knowledge_id_fkey" FOREIGN KEY (knowledge_id) REFERENCES knowledge(id) ON UPDATE CASCADE ON DELETE CASCADE not valid;
+
+alter table "public"."knowledge_brain" validate constraint "public_knowledge_brain_knowledge_id_fkey";
+
+alter table "public"."vectors" add constraint "public_vectors_knowledge_id_fkey" FOREIGN KEY (knowledge_id) REFERENCES knowledge(id) ON UPDATE CASCADE ON DELETE CASCADE not valid;
+
+alter table "public"."vectors" validate constraint "public_vectors_knowledge_id_fkey";
```
Comment (AmineDiro) [inline backend/supabase/migrations/20240814091741_notion-sync.sql:41]: This query should wait for data insertion before running validation on knowledge_id column

File: backend/supabase/migrations/20240814091741_notion-sync.sql [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+);
+
+
+alter table "public"."knowledge_brain" enable row level security;
+
+alter table "public"."vectors" drop column "file_sha1";
+
+alter table "public"."vectors" add column "knowledge_id" uuid;
+
+alter table "public"."vectors" alter column "id" set default gen_random_uuid();
```
Comment (AmineDiro) [inline backend/supabase/migrations/20240814091741_notion-sync.sql:19]: Nice 👍

File: docker-compose.dev.yml [hunk 6]
```diff
@@ -1,4 +1,4 @@
-version: "3.8"
+#version: "3.8"
```
Comment (AmineDiro) [inline docker-compose.dev.yml:1]: version comment

File: backend/worker/quivr_worker/celery_worker.py [hunk 7]
```diff
@@ -57,12 +59,15 @@
 brain_vectors = BrainsVectors()
 storage = Storage()
 notion_service: SyncNotionService | None = None
+knowledge_service: KnowledgeService | None = None
```
Comment (AmineDiro) [inline backend/worker/quivr_worker/celery_worker.py:62]: knowledge_service isn't used here

File: backend/worker/quivr_worker/celery_worker.py [hunk 8]
```diff
@@ -7,24 +7,26 @@
...
-    settings,
-)
+from quivr_api.models.settings import settings
 from quivr_api.modules.brain.integrations.Notion.Notion_connector import NotionConnector
 from quivr_api.modules.brain.repository.brains_vectors import BrainsVectors
 from quivr_api.modules.brain.service.brain_service import BrainService
 from quivr_api.modules.brain.service.brain_vector_service import BrainVectorService
+from quivr_api.modules.dependencies import get_supabase_client
 from quivr_api.modules.knowledge.repository.storage import Storage
+from quivr_api.modules.knowledge.service.knowledge_service import KnowledgeService
```
Comment (AmineDiro) [inline backend/worker/quivr_worker/celery_worker.py:17]: We can remove unused dep

File: backend/worker/quivr_worker/celery_worker.py [hunk 9]
```diff
@@ -91,18 +107,50 @@ def process_file_task(
...
+            notification_id=notification_id,
+            knowledge_id=knowledge_id,
+            source=source,
+            source_link=source_link,
+            delete_file=delete_file,
+        )
+    )
+
+
+async def aprocess_file_task(
```
Comment (AmineDiro) [inline backend/worker/quivr_worker/celery_worker.py:132]: To respect project structure we should move this function in `quivr_worker/process/porcess_file.py`

File: backend/worker/quivr_worker/celery_worker.py [hunk 10]
```diff
@@ -129,8 +176,12 @@ async def process_crawl_task(
...
         f"Task process_crawl_task started for url={crawl_website_url}, knowledge_id={knowledge_id}, brain_id={brain_id}, notification_id={notification_id}"
     )
-
-    brain_vector_service = BrainVectorService(brain_id)
+    global engine
+    assert engine
+    with Session(engine, expire_on_commit=False, autoflush=False) as session:
+        brain_vector_service = BrainVectorService(brain_id)
+        vector_repository = VectorRepository(session)
+        vector_service = VectorService(vector_repository)
```
Comment (AmineDiro) [inline backend/worker/quivr_worker/celery_worker.py:184]: Beware of context manager scope. The vector_service isn't in the scope for process_url_func

File: backend/worker/quivr_worker/celery_monitor.py [hunk 11]
```diff
@@ -100,10 +100,10 @@ async def handler_loop():
             )
 
 
-def notifier(app):
+async def notifier(app):
     state = app.events.State()
 
-    def handle_task_event(event):
+    async def handle_task_event(event):
```
Comment (AmineDiro) [inline backend/worker/quivr_worker/celery_monitor.py:106]: should be rolled back

File: backend/worker/quivr_worker/celery_monitor.py [hunk 12]
```diff
@@ -100,10 +100,10 @@ async def handler_loop():
             )
 
 
-def notifier(app):
+async def notifier(app):
```
Comment (AmineDiro) [inline backend/worker/quivr_worker/celery_monitor.py:103]: should be rolled back

File: backend/worker/quivr_worker/celery_worker.py [hunk 13]
```diff
@@ -129,8 +176,12 @@ async def process_crawl_task(
     logger.info(
         f"Task process_crawl_task started for url={crawl_website_url}, knowledge_id={knowledge_id}, brain_id={brain_id}, notification_id={notification_id}"
     )
-
-    brain_vector_service = BrainVectorService(brain_id)
+    global engine
+    assert engine
+    with Session(engine, expire_on_commit=False, autoflush=False) as session:
```
Comment (AmineDiro) [inline backend/worker/quivr_worker/celery_worker.py:181]: Is there any reason why this session is sync and not async ?

File: backend/worker/quivr_worker/process/process_file.py [hunk 14]
```diff
@@ -41,37 +41,32 @@ async def process_file(
...
         chunks=chunks,
-        document_vector_store=document_vector_store,
         brain_service=brain_service,
         brain_vector_service=brain_vector_service,
+        vector_service=vector_service,
     )
 
 
-def store_chunks(
+async def store_chunks(
```
Comment (AmineDiro) [inline backend/worker/quivr_worker/process/process_file.py:54]: No async code here but function is async

File: backend/api/quivr_api/vector/dto/input.py
Comment (AmineDiro) [inline backend/api/quivr_api/vector/dto/input.py:1]: The vector is now structured as a service. I think we can move it in `modules` 👍

File: backend/api/quivr_api/vector/dto/input.py [hunk 15]
```diff
@@ -0,0 +1,9 @@
+from uuid import UUID
+
+from pydantic import BaseModel
+
+
+class CreateVectorType(BaseModel):
+    content: str
+    metadata_: dict
```
Comment (AmineDiro) [inline backend/api/quivr_api/vector/dto/input.py:8]: fix underscore

File: backend/api/quivr_api/vector/entity/vector.py [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+        sa_column=Column(
+            PGUUID,
+            server_default=text("uuid_generate_v4()"),
+            primary_key=True,
+        ),
+    )
+    content: str = Field(default=None)
+    metadata_: dict = Field(default={}, sa_column=Column("metadata", JSON, default={}))
+    embedding: Optional[PGVector] = Field(
+        sa_column=Column(PGVector(1536))
```
Comment (AmineDiro) [inline backend/api/quivr_api/vector/entity/vector.py:25]: I think we should add an env variable in `.env` that is the embedding dimention

File: backend/api/quivr_api/vector/entity/vector.py [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+        arbitrary_types_allowed = True
+
+
+class VectorType(BaseModel):
+    id: UUID | None
+    content: str
+    metadata_: dict
+    knowledge_id: UUID
+
+class SimilaritySearchOutput(BaseModel):
```
Comment (AmineDiro) [inline backend/api/quivr_api/vector/entity/vector.py:39]: Thats nice !

File: backend/api/quivr_api/modules/dependencies.py [hunk 18]
```diff
@@ -56,24 +79,93 @@ def get_sync_session() -> Generator[Session, None, None]:
...
+def get_supabase_db() -> SupabaseDB:
+    global _supabase_db
+    if _supabase_db is None:
+        logger.info("Creating Supabase DB")
+        _supabase_db = SupabaseDB(get_supabase_client())
+    return _supabase_db
 
 
-def get_service(service: Type[S]) -> Callable[..., S]:
+def get_service(service: Type[S], asynchronous=True) -> Callable[..., S]:
```
Comment (AmineDiro) [inline backend/api/quivr_api/modules/dependencies.py:165]: add type hint ... This file should always type check

File: backend/api/quivr_api/modules/dependencies.py
Comment (AmineDiro) [inline backend/api/quivr_api/modules/dependencies.py:1]: Very subtle bug on `sync_engine` : The sync should not be future=True

File: backend/api/quivr_api/modules/sync/repository/sync_user.py [hunk 19]
```diff
@@ -20,6 +21,8 @@
 )
 
 notification_service = NotificationService()
+# knowledge_service = KnowledgeService()
+knowledge_service = get_service(KnowledgeService)()
```
Comment (AmineDiro) [inline backend/api/quivr_api/modules/sync/repository/sync_user.py:25]: How would this work ?

File: backend/api/quivr_api/modules/sync/repository/sync_repository.py [hunk 20]
```diff
@@ -18,6 +22,9 @@
 from quivr_api.modules.sync.repository.sync_interfaces import SyncInterface
 
 notification_service = NotificationService()
+# knowledge_service = KnowledgeService()
+knowledge_service = get_service(KnowledgeService)()
```
Comment (AmineDiro) [inline backend/api/quivr_api/modules/sync/repository/sync_repository.py:26]: How would this work ?

File: backend/api/quivr_api/modules/knowledge/repository/knowledges.py [hunk 21]
```diff
@@ -35,6 +35,13 @@ async def insert_knowledge(self, knowledge: KnowledgeDB) -> KnowledgeDB:
             if existing_knowledge:
```
Comment (AmineDiro) [inline backend/api/quivr_api/modules/knowledge/repository/knowledges.py:35]: This check shouldn't be on file_name an brain_id. (in later PR)
We should probably take an `upsert: bool` arg to upsert the knowledge if it exists and upsert is True

File: backend/api/quivr_api/modules/knowledge/controller/knowledge_routes.py [hunk 22]
```diff
@@ -39,6 +39,9 @@ async def list_knowledge_in_brain_endpoint(
     """
 
     validate_brain_authorization(brain_id=brain_id, user_id=current_user.id)
+    logger.debug(
```
Comment (AmineDiro) [inline backend/api/quivr_api/modules/knowledge/controller/knowledge_routes.py:42]: lowercase for logging.

We can remove this line as it was for internal debug not for production debug

**Review by AmineDiro — CHANGES_REQUESTED** [review]
Thank for the PR ! Some changes but overall very solid work and huge amount of changes !

File: backend/api/quivr_api/modules/dependencies.py (hunk 18 above)
Comment (chloedia) [reply backend/api/quivr_api/modules/dependencies.py:165]: it's already type hinted no ?

Comment (chloedia) [reply backend/api/quivr_api/modules/dependencies.py:165]: @AmineDiro

File: backend/api/quivr_api/modules/knowledge/repository/knowledges.py (hunk 21 above)
Comment (chloedia) [reply backend/api/quivr_api/modules/knowledge/repository/knowledges.py:35]: the check is done before hand, maybe not check twice ? though, if the file_name exists maybe suggest a new name ?

File: backend/api/quivr_api/vector/repository/vectors_repository.py [hunk 23]
```diff
@@ -0,0 +1,115 @@
...
+from quivr_api.modules.dependencies import BaseRepository
+from quivr_api.vector.entity.vector import SimilaritySearchOutput, Vector
+
+logger = get_logger(__name__)
+
+
+class VectorRepository(BaseRepository):
+    def __init__(self, session: Session):
+        super().__init__(session)
+        self.session = session
```
[the remark this answers is not in the document]
Comment (chloedia) [reply backend/api/quivr_api/vector/repository/vectors_repository.py:17]: we need it sync because of the similarity search function

[the remark this answers is not in the document]
Comment (chloedia) [reply backend/api/quivr_api/vector/repository/vectors_repository.py:17]: I am not sure that i understand the comment

File: backend/api/quivr_api/vector/service/vector_service.py [hunk 24]
```diff
@@ -0,0 +1,70 @@
...
+                content=chunk.page_content,
+                metadata_=chunk.metadata,
+                embedding=embeddings[i], #type: ignore
+                knowledge_id=knowledge_id,
+            )
+            for i, chunk in enumerate(chunks)
+        ]
+        created_vector = self.repository.create_vectors(new_vectors)
+
+        return [str(vector.id) for vector in created_vector]
```
[the remark this answers is not in the document]
Comment (chloedia) [reply backend/api/quivr_api/vector/service/vector_service.py:41]: Didn't you say we need to return only the ids of the created vectors ? @AmineDiro

File: backend/worker/quivr_worker/celery_worker.py (hunk 13 above)
Comment (chloedia) [reply backend/worker/quivr_worker/celery_worker.py:181]: because vector service is sync

File: backend/core/quivr_core/models.py (hunk 2 above)
Comment (chloedia) [reply backend/core/quivr_core/models.py:86]: ?? it is exactly the same as in release/quivr-core

File: backend/api/quivr_api/vector/service/vector_service.py [hunk 25]
```diff
@@ -0,0 +1,70 @@
...
+        )
+        # FIXME ADD a check in case of failure
+        embeddings = self._embedding.embed_documents(
+            [chunk.page_content for chunk in chunks]
+        )
+        new_vectors = [
+            Vector(
+                content=chunk.page_content,
+                metadata_=chunk.metadata,
+                embedding=embeddings[i], #type: ignore
```
Comment (AmineDiro) [inline backend/api/quivr_api/vector/service/vector_service.py:34]: Are we sure the API returns the embeddings in the correct order ?

File: backend/api/quivr_api/models/settings.py [hunk 26]
```diff
@@ -125,6 +121,7 @@ class BrainSettings(BaseSettings):
     langfuse_secret_key: str | None = None
     pg_database_url: str
     pg_database_async_url: str
+    embedding_dim: int = int(os.getenv("EMBEDDING_DIM", 1536))
```
Comment (AmineDiro) [inline backend/api/quivr_api/models/settings.py:124]: The BrainSettings in a SettingsConfiDict. It automatically resolves values from .env and environement variable:
https://docs.pydantic.dev/latest/concepts/pydantic_settings/


So no need to set it here 👍

File: backend/worker/quivr_worker/celery_monitor.py [hunk 27]
```diff
@@ -9,13 +9,16 @@
...
 from quivr_api.logger import get_logger
 from quivr_api.modules.dependencies import async_engine
-from quivr_api.modules.knowledge.repository.knowledges import KnowledgeRepository
-from quivr_api.modules.knowledge.service.knowledge_service import KnowledgeService
-from quivr_api.modules.notification.dto.inputs import NotificationUpdatableProperties
-from quivr_api.modules.notification.entity.notification import NotificationsStatusEnum
-from quivr_api.modules.notification.service.notification_service import (
-    NotificationService,
-)
+from quivr_api.modules.knowledge.repository.knowledges import \
```
Comment (AmineDiro) [inline backend/worker/quivr_worker/celery_monitor.py:12]: Check formatting configuration after running pre-commit 🤔

**Review by AmineDiro — APPROVED** [review]
Thx for the work !
